File: frontend/src/plugins/UserService/APIs/ModifyUserStatusMessage.ts
/**
 * ModifyUserStatusMessage
 * desc: 管理员通过提供userID和新的封禁天数，更新用户的封禁字段。
 * @param adminToken: String (管理员身份验证的Token)
 * @param userID: String (需要被修改状态的用户ID)
 * @param banDays: Int (新的封禁天数)
 * @return result: String (修改状态的结果信息 (如: '用户状态修改成功！'))
 */
import { TongWenMessage } from '../../TongWenAPI/TongWenMessage'



export class ModifyUserStatusMessage extends TongWenMessage {
    constructor(
        public adminToken: string,
        public userID: string,
        public banDays: number
    ) {
        super()
    }
    getAddress(): string {
        return "127.0.0.1:10010"
    }
}

File: UserService/src/main/scala/APIs/UserService/ModifyUserStatusMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID


/**
 * ModifyUserStatusMessage
 * desc: 通过提供 userID 和新的封禁天数，更新用户的封禁字段。
 * @param userID: String (需要被修改状态的用户ID)
 * @param banDays: Int (新的封禁天数)
 * @return result: String (修改状态的结果信息 (如: '用户状态修改成功！'))
 */

case class ModifyUserStatusMessage(
  userID: String,
  banDays: Int
) extends API[String](UserServiceCode)



case object ModifyUserStatusMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[ModifyUserStatusMessage] = deriveEncoder
  private val circeDecoder: Decoder[ModifyUserStatusMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[ModifyUserStatusMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[ModifyUserStatusMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[ModifyUserStatusMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given modifyUserStatusMessageEncoder: Encoder[ModifyUserStatusMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given modifyUserStatusMessageDecoder: Decoder[ModifyUserStatusMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }


}

